File: file-handling-io/file_handling_io.py
import csv
import json
from pathlib import Path


def main() -> None:
    print("Language: Python")

    base_dir = Path(__file__).resolve().parent
    data_dir = base_dir / "data"
    data_dir.mkdir(exist_ok=True)

    text_path = data_dir / "sample.txt"
    json_path = data_dir / "sample.json"
    csv_path = data_dir / "sample.csv"

    # Write text with buffering and newline handling.
    with text_path.open("w", encoding="utf-8", buffering=8192) as f:
        f.write("line-1\n")
        f.write("line-2\n")

    # Read text back.
    with text_path.open("r", encoding="utf-8") as f:
        text_content = f.read()
    print("Text read:")
    print(text_content.strip())

    # JSON write/read.
    payload = {
        "name": "Nazmul",
        "skills": ["python", "javascript", "java", "csharp", "cpp"],
        "active": True
    }
    with json_path.open("w", encoding="utf-8") as f:
        json.dump(payload, f, indent=2)
    with json_path.open("r", encoding="utf-8") as f:
        parsed_json = json.load(f)
    print("JSON parsed:", parsed_json)

    # CSV write/read.
    rows = [
        ["id", "name", "score"],
        ["1", "Alice", "95"],
        ["2", "Bob", "88"]
    ]
    with csv_path.open("w", encoding="utf-8", newline="") as f:
        writer = csv.writer(f)
        writer.writerows(rows)

    with csv_path.open("r", encoding="utf-8", newline="") as f:
        reader = csv.DictReader(f)
        parsed_csv = list(reader)
    print("CSV parsed:", parsed_csv)

    # Path API examples.
    print("Path exists:", text_path.exists())
    print("Path name:", text_path.name)
    print("Parent dir:", text_path.parent)


if __name__ == "__main__":
    main()

File: file-handling-io/fileHandlingIo.js
const fs = require("fs");
const fsp = require("fs/promises");
const path = require("path");

async function main() {
  console.log("Language: JavaScript");

  const baseDir = __dirname;
  const dataDir = path.join(baseDir, "data");
  await fsp.mkdir(dataDir, { recursive: true });

  const textPath = path.join(dataDir, "sample.txt");
  const jsonPath = path.join(dataDir, "sample.json");
  const csvPath = path.join(dataDir, "sample.csv");

  // Write text using a buffered stream.
  await new Promise((resolve, reject) => {
    const stream = fs.createWriteStream(textPath, { encoding: "utf8", highWaterMark: 16 * 1024 });
    stream.on("error", reject);
    stream.on("finish", resolve);
    stream.write("line-1\n");
    stream.write("line-2\n");
    stream.end();
  });

  // Read text back.
  const textContent = await fsp.readFile(textPath, "utf8");
  console.log("Text read:\n" + textContent.trim());

  // JSON write/read.
  const payload = {
    name: "Nazmul",
    skills: ["python", "javascript", "java", "csharp", "cpp"],
    active: true,
  };
  await fsp.writeFile(jsonPath, JSON.stringify(payload, null, 2), "utf8");
  const parsedJson = JSON.parse(await fsp.readFile(jsonPath, "utf8"));
  console.log("JSON parsed:", parsedJson);

  // CSV write/read (manual parsing for comparison purpose).
  const csvRows = [
    "id,name,score",
    "1,Alice,95",
    "2,Bob,88",
  ];
  await fsp.writeFile(csvPath, csvRows.join("\n") + "\n", "utf8");

  const csvContent = await fsp.readFile(csvPath, "utf8");
  const [headerLine, ...dataLines] = csvContent.trim().split("\n");
  const headers = headerLine.split(",");
  const parsedCsv = dataLines.map((line) => {
    const parts = line.split(",");
    return Object.fromEntries(headers.map((h, i) => [h, parts[i]]));
  });
  console.log("CSV parsed:", parsedCsv);

  // Path API examples.
  console.log("Path exists:", fs.existsSync(textPath));
  console.log("Path basename:", path.basename(textPath));
  console.log("Path dirname:", path.dirname(textPath));
}

main().catch((err) => {
  console.error("Error:", err);
  process.exit(1);
});
